modyn/supervisor/internal/triggers/datadrifttrigger.py

```diff
@@ -0,0 +1,333 @@
+                trigger_idx = len(new_data) - (cached_data_points - trigger_data_points) - 1
+
+                logger.debug(
+                    f"[DataDriftDetector][Prev Trigger {self.previous_trigger_id}]"
+                    + f"[Dataset {self.dataloader_info.dataset_id}]"
+                    + f"[Trigger data points] {trigger_data_points}, [Trigger index] {trigger_idx}"
+                )
+                _, timestamps, _ = zip(
+                    *self.data_cache[detection_data_idx_start:detection_data_idx_end]
+                )  # type: ignore
+                num_per_t = {}
+                for t in timestamps:
+                    num_per_t[t] = num_per_t.get(t, 0) + 1
+                logger.debug(
+                    f"[DataDriftDetector][Prev Trigger {self.previous_trigger_id}]"
+                    + f"[Dataset {self.dataloader_info.dataset_id}]"
+                    + f"[Trigger Timestamps] {len(num_per_t)}:{num_per_t}"
+                )
+
+                # update bookkeeping and pointers
+                cached_data_points -= trigger_data_points
+                detection_data_idx_start = detection_data_idx_end
+                yield trigger_idx
+
+        # remove triggered data
+        self.data_cache = self.data_cache[detection_data_idx_start:]
+        self.untriggered_detection_data_points = detection_data_idx_end - detection_data_idx_start
+
+    def _init_dataloader_info(self) -> None:
+        assert self.pipeline_id is not None
+        assert self.pipeline_config is not None
+        assert self.modyn_config is not None
+        assert self.base_dir is not None
+
+        training_config = self.pipeline_config["training"]
+        data_config = self.pipeline_config["data"]
+
+        if "num_prefetched_partitions" in training_config:
+            num_prefetched_partitions = training_config["num_prefetched_partitions"]
+        else:
+            if "prefetched_partitions" in training_config:
+                raise ValueError(
+                    "Found `prefetched_partitions` instead of `num_prefetched_partitions`in training configuration."
+                    + " Please rename/remove that configuration"
+                )
+            logger.warning("Number of prefetched partitions not explicitly given in training config - defaulting to 1.")
+            num_prefetched_partitions = 1
+
+        if "parallel_prefetch_requests" in training_config:
+            parallel_prefetch_requests = training_config["parallel_prefetch_requests"]
+        else:
+            logger.warning(
+                "Number of parallel prefetch requests not explicitly given in training config - defaulting to 1."
+            )
+            parallel_prefetch_requests = 1
+
+        if "tokenizer" in data_config:
+            tokenizer = data_config["tokenizer"]
+        else:
+            tokenizer = None
+
+        if "transformations" in data_config:
+            transform_list = data_config["transformations"]
+        else:
```

Comment:
I have seen this multiple times in the repo. Can we somehow unify this?

Reply:
TODO: next PR